tools/integration-test/src/tests/manual/client_expiration.rs

```diff
@@ -0,0 +1,261 @@
+        _config: &SharedConfig,
+        _registry: &SharedRegistry<impl ChainHandle>,
+    ) -> Result<Option<SupervisorHandle>, Error> {
+        Ok(None)
+    }
+}
+
+impl HasOverrides for ClientExpirationTest {
+    type Overrides = ExpirationTestOverrides;
+
+    fn get_overrides(&self) -> &ExpirationTestOverrides {
+        &ExpirationTestOverrides
+    }
+}
+
+impl HasOverrides for ConnectionExpirationTest {
+    type Overrides = ExpirationTestOverrides;
+
+    fn get_overrides(&self) -> &ExpirationTestOverrides {
+        &ExpirationTestOverrides
+    }
+}
+
+impl HasOverrides for ChannelExpirationTest {
+    type Overrides = ExpirationTestOverrides;
+
+    fn get_overrides(&self) -> &ExpirationTestOverrides {
+        &ExpirationTestOverrides
+    }
+}
+
+impl BinaryChainTest for ConnectionExpirationTest {
+    fn run<ChainA: ChainHandle, ChainB: ChainHandle>(
+        &self,
+        _config: &TestConfig,
+        chains: ConnectedChains<ChainA, ChainB>,
+    ) -> Result<(), Error> {
+        // We spawn a supervisor first, but the refresh client is not
+        // triggered because it is currently tied to channels.
+        let supervisor =
+            spawn_supervisor(chains.config.clone(), chains.registry.clone(), None, false)?;
+
+        wait_for_client_expiry();
+
+        info!("Trying to create connection after client is expired");
+
+        // We can submit an CONN_INIT packet and the connection worker will
+        // pick it up, and we want to observe that it abort due to frozen client
+        // rather than looping indefinitely.
+
+        init_connection(
+            &chains.handle_a,
+            &chains.handle_b,
+            &chains.client_b_to_a.tagged_client_id(),
+            &chains.client_a_to_b.tagged_client_id(),
+        )?;
+
+        // New work in progress
+
+        sleep(Duration::from_secs(10));
+
+        info!("Trying to new connection and worker after previous connection worker failed");
+
+        let client_b_to_a_2 = bootstrap_foreign_client(&chains.handle_b, &chains.handle_a)?;
+
+        let client_a_to_b_2 = bootstrap_foreign_client(&chains.handle_a, &chains.handle_b)?;
+
+        // Problem: if the connection worker for a chain abort due to frozen client,
+        // it won't handle subsequent init connection packets.
+
+        init_connection(
+            &chains.handle_a,
+            &chains.handle_b,
+            &client_b_to_a_2.tagged_client_id(),
+            &client_a_to_b_2.tagged_client_id(),
+        )?;
+
+        sleep(Duration::from_secs(10));
+
+        supervisor.shutdown();
+
+        // Even if we respawn the supervisor, it would still abort immediately
+        // due to the first frozen client error. We are effectively stuck at
+        // either having to keep retrying on frozen clients or not able to
+        // handle any other new connection due to frozen clients.
+
+        let _supervisor = spawn_supervisor(chains.config.clone(), chains.registry, None, false)?;
```

Comment:
> that the connection and channel workers are configured to run with one worker per chain, rather than one worker per IBC client.

Can you expand on this? I don't understand what you mean exactly. Currently there is one worker per connection (or per channel); there isn't a worker per chain. Do you mean that each connection (and channel..) worker should be parametrized with the underlying IBC client?

Reply:
Hmm you are right. It was probably my oversight when observing the logs at the end of the day. I have run the tests again and confirmed that the connection handshake worker in an unexpired client is not interrupted the other expired clients.